// node-opcua-server/dist/base_server.js
"use strict";
/**
 * @module node-opcua-server
 */
// tslint:disable:no-console
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPCUABaseServer = void 0;
const node_fs_1 = __importDefault(require("node:fs"));
const node_net_1 = require("node:net");
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
const global_mutex_1 = require("@ster5/global-mutex");
const chalk_1 = __importDefault(require("chalk"));
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_certificate_manager_1 = require("node-opcua-certificate-manager");
const node_opcua_client_1 = require("node-opcua-client");
const node_opcua_common_1 = require("node-opcua-common");
const web_1 = require("node-opcua-crypto/web");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_date_time_1 = require("node-opcua-date-time");
const node_opcua_debug_1 = require("node-opcua-debug");
const node_opcua_hostname_1 = require("node-opcua-hostname");
const node_opcua_service_discovery_1 = require("node-opcua-service-discovery");
const node_opcua_service_endpoints_1 = require("node-opcua-service-endpoints");
const node_opcua_service_secure_channel_1 = require("node-opcua-service-secure-channel");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const node_opcua_utils_1 = require("node-opcua-utils");
const doDebug = (0, node_opcua_debug_1.checkDebugFlag)(__filename);
const debugLog = (0, node_opcua_debug_1.make_debugLog)(__filename);
const errorLog = (0, node_opcua_debug_1.make_errorLog)(__filename);
const warningLog = (0, node_opcua_debug_1.make_warningLog)(__filename);
const default_server_info = {
    // The globally unique identifier for the application instance. This URI is used as
    // ServerUri in Services if the application is a Server.
    applicationUri: (0, node_opcua_common_1.makeApplicationUrn)(node_os_1.default.hostname(), "NodeOPCUA-Server"),
    // The globally unique identifier for the product.
    productUri: "NodeOPCUA-Server",
    // A localized descriptive name for the application.
    applicationName: { text: "NodeOPCUA", locale: "en" },
    applicationType: node_opcua_service_endpoints_1.ApplicationType.Server,
    gatewayServerUri: "",
    discoveryProfileUri: "",
    discoveryUrls: []
};
function cleanupEndpoint(endpoint) {
    if (endpoint._on_new_channel) {
        (0, node_opcua_assert_1.assert)(typeof endpoint._on_new_channel === "function");
        endpoint.removeListener("newChannel", endpoint._on_new_channel);
        endpoint._on_new_channel = undefined;
    }
    if (endpoint._on_close_channel) {
        (0, node_opcua_assert_1.assert)(typeof endpoint._on_close_channel === "function");
        endpoint.removeListener("closeChannel", endpoint._on_close_channel);
        endpoint._on_close_channel = undefined;
    }
    if (endpoint._on_connectionRefused) {
        (0, node_opcua_assert_1.assert)(typeof endpoint._on_connectionRefused === "function");
        endpoint.removeListener("connectionRefused", endpoint._on_connectionRefused);
        endpoint._on_connectionRefused = undefined;
    }
    if (endpoint._on_openSecureChannelFailure) {
        (0, node_opcua_assert_1.assert)(typeof endpoint._on_openSecureChannelFailure === "function");
        endpoint.removeListener("openSecureChannelFailure", endpoint._on_openSecureChannelFailure);
        endpoint._on_openSecureChannelFailure = undefined;
    }
    if (endpoint._on_channel_secured) {
        (0, node_opcua_assert_1.assert)(typeof endpoint._on_channel_secured === "function");
        endpoint.removeListener("channelSecured", endpoint._on_channel_secured);
        endpoint._on_channel_secured = undefined;
    }
}
const emptyCallback = () => {
    /* empty */
};
// biome-ignore lint/suspicious/noExplicitAny: must propagate EventEmitter generic
class OPCUABaseServer extends node_opcua_common_1.OPCUASecureObject {
    static makeServiceFault = makeServiceFault;
    /**
     * The type of server
     */
    get serverType() {
        return this.serverInfo.applicationType;
    }
    serverInfo;
    endpoints;
    serverCertificateManager;
    capabilitiesForMDNS;
    _preInitTask;
    options;
    constructor(options) {
        options = options || {};
        if (!options.serverCertificateManager) {
            options.serverCertificateManager = (0, node_opcua_certificate_manager_1.getDefaultCertificateManager)("PKI");
        }
        options.privateKeyFile = options.privateKeyFile || options.serverCertificateManager.privateKey;
        options.certificateFile =
            options.certificateFile || node_path_1.default.join(options.serverCertificateManager.rootDir, "own/certs/certificate.pem");
        super(options);
        this.serverCertificateManager = options.serverCertificateManager;
        this.capabilitiesForMDNS = [];
        this.endpoints = [];
        this.options = options;
        this._preInitTask = [];
        const serverInfo = {
            ...default_server_info,
            ...options.serverInfo
        };
        serverInfo.applicationName = (0, node_opcua_data_model_1.coerceLocalizedText)(serverInfo.applicationName);
        this.serverInfo = new node_opcua_service_endpoints_1.ApplicationDescription(serverInfo);
        if (this.serverInfo.applicationName.toString().match(/urn:/)) {
            errorLog("[NODE-OPCUA-E06] application name cannot be a urn", this.serverInfo.applicationName.toString());
        }
        this.serverInfo.applicationName.locale = this.serverInfo.applicationName.locale || "en";
        if (!this.serverInfo.applicationName.locale) {
            warningLog("[NODE-OPCUA-W24] the server applicationName must have a valid locale : ", this.serverInfo.applicationName.toString());
        }
        const __applicationUri = serverInfo.applicationUri || "";
        Object.defineProperty(this.serverInfo, "applicationUri", {
            get: () => (0, node_opcua_hostname_1.resolveFullyQualifiedDomainName)(__applicationUri),
            configurable: true
        });
        this._preInitTask.push(async () => {
            await (0, node_opcua_hostname_1.extractFullyQualifiedDomainName)();
        });
        this._preInitTask.push(async () => {
            await this.initializeCM();
        });
    }
    /**
     * Return additional DNS hostnames to include in the self-signed
     * certificate's SubjectAlternativeName (SAN).
     *
     * The base implementation returns an empty array. Subclasses
     * (e.g. `OPCUAServer`) override this to include hostnames from
     * `alternateHostname` and `advertisedEndpoints`.
     *
     * @internal
     */
    getConfiguredHostnames() {
        return [];
    }
    /**
     * Return additional IP addresses to include in the self-signed
     * certificate's SubjectAlternativeName (SAN) iPAddress entries.
     *
     * The base implementation returns an empty array. Subclasses
     * (e.g. `OPCUAServer`) override this to include IP literals
     * found in `alternateHostname` and `advertisedEndpoints`.
     *
     * These IPs are considered **explicitly configured** by the
     * user and are therefore checked by `checkCertificateSAN()`.
     * In contrast, auto-detected IPs from `getIpAddresses()` are
     * included in the certificate at creation time but are NOT
     * checked later — see `checkCertificateSAN()` for rationale.
     *
     * @internal
     */
    getConfiguredIPs() {
        return [];
    }
    async createDefaultCertificate() {
        if (node_fs_1.default.existsSync(this.certificateFile)) {
            return;
        }
        if (!(0, node_opcua_utils_1.checkFileExistsAndIsNotEmpty)(this.certificateFile)) {
            await (0, global_mutex_1.withLock)({ fileToLock: `${this.certificateFile}.mutex` }, async () => {
                if ((0, node_opcua_utils_1.checkFileExistsAndIsNotEmpty)(this.certificateFile)) {
                    return;
                }
                const applicationUri = this.serverInfo.applicationUri || "<missing application uri>";
                const fqdn = (0, node_opcua_hostname_1.getFullyQualifiedDomainName)();
                const hostname = (0, node_opcua_hostname_1.getHostname)();
                const dns = [...new Set([fqdn, hostname, ...this.getConfiguredHostnames()])].sort();
                // Include both auto-detected IPs and explicitly configured IPs.
                // Auto-detected IPs (getIpAddresses) are ephemeral — they depend on
                // the current network state (WiFi, tethering, VPN, roaming) and may
                // change between reboots. They are included here so that the initial
                // certificate covers the current network configuration, but they are
                // NOT checked by checkCertificateSAN() to avoid noisy warnings when
                // the network changes. Only explicitly configured IPs (from
                // alternateHostname / advertisedEndpoints) are checked at startup.
                const ip = [...new Set([...(0, node_opcua_hostname_1.getIpAddresses)(), ...this.getConfiguredIPs()])].sort();
                await this.serverCertificateManager.createSelfSignedCertificate({
                    applicationUri,
                    dns,
                    ip,
                    outputFile: this.certificateFile,
                    subject: (0, node_opcua_certificate_manager_1.makeSubject)(this.serverInfo.applicationName.text || "<missing application name>", hostname),
                    startDate: new Date(),
                    validity: 365 * 10 // 10 years
                });
            });
        }
    }
    async initializeCM() {
        await this.serverCertificateManager.initialize();
        await this.createDefaultCertificate();
        debugLog("privateKey      = ", this.privateKeyFile, this.serverCertificateManager.privateKey);
        debugLog("certificateFile = ", this.certificateFile);
        this._checkCertificateSanMismatch();
        await (0, node_opcua_client_1.performCertificateSanityCheck)(this, "server", this.serverCertificateManager, this.serverInfo.applicationUri || "");
    }
    /**
     * Compare the current certificate's SAN entries against all
     * explicitly configured hostnames and IPs, and return any
     * that are missing.
     *
     * Returns an empty array when the certificate covers every
     * configured hostname and IP.
     *
     * **Important — ephemeral IP mitigation:**
     * Auto-detected IPs (from `getIpAddresses()`) are deliberately
     * NOT included in this check. Network interfaces are transient
     * — WiFi IPs change on reconnect, tethering IPs appear/disappear,
     * VPN adapters come and go. Including them would cause the
     * `[NODE-OPCUA-W26]` warning to fire on every server restart
     * whenever the network state differs from when the certificate
     * was originally created.
     *
     * Only **explicitly configured** values are checked:
     * - Hostnames: FQDN, os.hostname(), `alternateHostname` (non-IP),
     *   hostnames from `advertisedEndpoints` URLs
     * - IPs: IP literals from `alternateHostname`, IP literals
     *   from `advertisedEndpoints` URLs
     *
     * The certificate itself still includes auto-detected IPs at
     * creation time — this is fine because it captures the network
     * state at that moment. But the *mismatch warning* only fires
     * for things the user explicitly asked for.
     */
    checkCertificateSAN() {
        const certDer = this.getCertificate();
        const info = (0, web_1.exploreCertificate)(certDer);
        const sanDns = info.tbsCertificate.extensions?.subjectAltName?.dNSName || [];
        const sanIpsHex = info.tbsCertificate.extensions?.subjectAltName?.iPAddress || [];
        const fqdn = (0, node_opcua_hostname_1.getFullyQualifiedDomainName)();
        const hostname = (0, node_opcua_hostname_1.getHostname)();
        const expectedDns = [...new Set([fqdn, hostname, ...this.getConfiguredHostnames()])].sort();
        // Only check explicitly configured IPs — NOT auto-detected ones.
        // See JSDoc above for the rationale (ephemeral network interfaces).
        const expectedIps = [...new Set(this.getConfiguredIPs())].sort();
        const missingDns = expectedDns.filter((name) => !sanDns.includes(name));
        // exploreCertificate returns iPAddress entries as hex strings
        // Only IPv4 addresses can be converted with ipv4ToHex here; IPv6 (and invalid) IPs are skipped.
        const missingIps = expectedIps.filter((ip) => {
            const family = (0, node_net_1.isIP)(ip);
            if (family === 4) {
                return !sanIpsHex.includes((0, node_opcua_hostname_1.ipv4ToHex)(ip));
            }
            // IPv6 or invalid literals are currently not matched against SAN iPAddress entries here.
            return false;
        });
        return [...missingDns, ...missingIps];
    }
    /**
     * Delete the existing self-signed certificate and create a new
     * one that includes all currently configured hostnames.
     *
     * @throws if the current certificate was NOT self-signed
     *         (i.e. issued by a CA or GDS)
     */
    async regenerateSelfSignedCertificate() {
        // guard: only allow regeneration of self-signed certs
        const certDer = this.getCertificate();
        const info = (0, web_1.exploreCertificate)(certDer);
        const issuer = info.tbsCertificate.issuer;
        const subject = info.tbsCertificate.subject;
        const isSelfSigned = issuer.commonName === subject.commonName && issuer.organizationName === subject.organizationName;
        if (!isSelfSigned) {
            throw new Error("Cannot regenerate certificate: current certificate is not self-signed (issued by a CA or GDS)");
        }
        // delete old cert
        if (node_fs_1.default.existsSync(this.certificateFile)) {
            node_fs_1.default.unlinkSync(this.certificateFile);
        }
        // recreate with current hostnames
        await this.createDefaultCertificate();
        // invalidate cached cert so next getCertificate() reloads from disk
        (0, node_opcua_common_1.invalidateCachedSecrets)(this);
    }
    _checkCertificateSanMismatch() {
        try {
            const missing = this.checkCertificateSAN();
            if (missing.length > 0) {
                warningLog(`[NODE-OPCUA-W26] Certificate SAN is missing the following configured hostnames/IPs: ${missing.join(", ")}. ` +
                    "Clients with strict certificate validation may reject connections for these entries. " +
                    "Use server.regenerateSelfSignedCertificate() to fix this.");
            }
        }
        catch (_err) {
            // ignore errors during SAN check (e.g. cert not yet loaded)
        }
    }
    start(...args) {
        const callback = args[0];
        if (!callback || args.length === 0) {
            return this.startAsync();
        }
        else {
            this.startAsync()
                .then(() => {
                callback();
            })
                .catch((err) => callback(err));
        }
    }
    async performPreInitialization() {
        const tasks = this._preInitTask;
        this._preInitTask = [];
        for (const task of tasks) {
            await task();
        }
    }
    async startAsync() {
        await this.performPreInitialization();
        (0, node_opcua_assert_1.assert)(Array.isArray(this.endpoints));
        (0, node_opcua_assert_1.assert)(this.endpoints.length > 0, "We need at least one end point");
        (0, node_opcua_date_time_1.installPeriodicClockAdjustment)();
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const server = this;
        const _on_new_channel = function (channel) {
            server.emit("newChannel", channel, this);
        };
        const _on_channel_secured = function (channel) {
            // Install a response interceptor once per channel so the
            // server can emit "response" events for diagnostics.
            // Done here (after OpenSecureChannel) rather than at
            // newChannel time, so the interceptor can rely on
            // securityPolicy/securityMode being populated.
            channel.setResponseInterceptor((_msg, response1) => {
                server.emit("response", response1, channel);
            });
            server.emit("channelSecured", channel, this);
        };
        const _on_close_channel = function (channel) {
            server.emit("closeChannel", channel, this);
        };
        const _on_connectionRefused = function (socketData) {
            server.emit("connectionRefused", socketData, this);
        };
        const _on_openSecureChannelFailure = function (socketData, channelData) {
            server.emit("openSecureChannelFailure", socketData, channelData, this);
        };
        const promises = [];
        for (const endpoint of this.endpoints) {
            (0, node_opcua_assert_1.assert)(!endpoint._on_close_channel);
            endpoint._on_new_channel = _on_new_channel;
            endpoint.on("newChannel", endpoint._on_new_channel);
            endpoint._on_channel_secured = _on_channel_secured;
            endpoint.on("channelSecured", endpoint._on_channel_secured);
            endpoint._on_close_channel = _on_close_channel;
            endpoint.on("closeChannel", endpoint._on_close_channel);
            endpoint._on_connectionRefused = _on_connectionRefused;
            endpoint.on("connectionRefused", endpoint._on_connectionRefused);
            endpoint._on_openSecureChannelFailure = _on_openSecureChannelFailure;
            endpoint.on("openSecureChannelFailure", endpoint._on_openSecureChannelFailure);
            promises.push(new Promise((resolve, reject) => endpoint.start((err) => (err ? reject(err) : resolve()))));
        }
        await Promise.all(promises);
    }
    /**
     * shutdown all server endPoints
     */
    shutdown(done) {
        (0, node_opcua_assert_1.assert)(typeof done === "function");
        (0, node_opcua_date_time_1.uninstallPeriodicClockAdjustment)();
        this.serverCertificateManager.dispose().then(() => {
            debugLog("OPCUABaseServer#shutdown starting");
            const promises = this.endpoints.map((endpoint) => {
                return new Promise((resolve, reject) => {
                    cleanupEndpoint(endpoint);
                    endpoint.shutdown((err) => (err ? reject(err) : resolve()));
                });
            });
            Promise.all(promises)
                .then(() => {
                debugLog("shutdown completed");
                done();
            })
                .catch((err) => done(err));
        });
    }
    shutdownChannels(callback) {
        (0, node_opcua_assert_1.assert)(typeof callback === "function");
        // c8 ignore next
        if (!callback)
            throw new Error("thenify is not available");
        debugLog("OPCUABaseServer#shutdownChannels");
        const promises = this.endpoints.map((endpoint) => {
            return new Promise((resolve, reject) => {
                debugLog(" shutting down endpoint ", endpoint.endpointDescriptions()[0].endpointUrl);
                endpoint.abruptlyInterruptChannels();
                endpoint.shutdown((err) => (err ? reject(err) : resolve()));
            });
        });
        Promise.all(promises)
            .then(() => callback())
            .catch((err) => callback?.(err));
    }
    /**
     * @private
     */
    on_request(message, channel) {
        (0, node_opcua_assert_1.assert)(message.request);
        (0, node_opcua_assert_1.assert)(message.requestId !== 0);
        const request = message.request;
        // prepare request
        this.prepare(message, channel);
        if (doDebug) {
            debugLog(chalk_1.default.green.bold("--------------------------------------------------------"), channel.channelId, request.schema.name);
        }
        let errMessage;
        let response;
        this.emit("request", request, channel);
        try {
            // handler must be named _on_ActionRequest()
            const handler = this[`_on_${request.schema.name}`];
            if (typeof handler === "function") {
                handler.call(this, message, channel);
            }
            else {
                errMessage = `[NODE-OPCUA-W07] Unsupported Service : ${request.schema.name}`;
                warningLog(errMessage);
                debugLog(chalk_1.default.red.bold(errMessage));
                response = makeServiceFault(node_opcua_status_code_1.StatusCodes.BadServiceUnsupported, [errMessage]);
                channel.send_response("MSG", response, message, emptyCallback);
            }
        }
        catch (err) {
            /* c8 ignore next */
            const errMessage1 = `[NODE-OPCUA-W08] EXCEPTION CAUGHT WHILE PROCESSING REQUEST !! ${request.schema.name}`;
            warningLog(chalk_1.default.red.bold(errMessage1));
            warningLog(request.toString());
            (0, node_opcua_debug_1.displayTraceFromThisProjectOnly)(err);
            let additional_messages = [];
            additional_messages.push(`EXCEPTION CAUGHT WHILE PROCESSING REQUEST !!! ${request.schema.name}`);
            if (err instanceof Error) {
                additional_messages.push(err.message);
                if (err.stack) {
                    additional_messages = additional_messages.concat(err.stack.split("\n"));
                }
            }
            response = makeServiceFault(node_opcua_status_code_1.StatusCodes.BadInternalError, additional_messages);
            channel.send_response("MSG", response, message, emptyCallback);
        }
    }
    /**
     * Find endpoint descriptions matching a given endpoint URL.
     *
     * When `endpointUrl` is provided, only endpoints whose URL matches
     * (case-insensitive) are returned. When `null` or omitted, all
     * endpoints from every `OPCUAServerEndPoint` are returned.
     *
     * This is the shared resolution path used by both `GetEndpoints`
     * and `CreateSession` (`validate_security_endpoint`).
     *
     * @internal (was _get_endpoints)
     */
    findMatchingEndpoints(endpointUrl) {
        let endpoints = [];
        for (const endPoint of this.endpoints) {
            const ep = endPoint.endpointDescriptions();
            const epFiltered = endpointUrl ? ep.filter((e) => (0, node_opcua_utils_1.matchUri)(e.endpointUrl, endpointUrl)) : ep;
            endpoints = endpoints.concat(epFiltered);
        }
        return endpoints;
    }
    /**
     * get one of the possible endpointUrl
     */
    getEndpointUrl() {
        return this.findMatchingEndpoints()[0].endpointUrl || "";
    }
    getDiscoveryUrls() {
        const discoveryUrls = this.endpoints.map((e) => {
            return e.endpointDescriptions()[0].endpointUrl || "";
        });
        return discoveryUrls;
    }
    getServers(_channel) {
        this.serverInfo.discoveryUrls = this.getDiscoveryUrls();
        const servers = [this.serverInfo];
        return servers;
    }
    suspendEndPoints(callback) {
        /* c8 ignore next */
        if (!callback) {
            throw new Error("Internal Error");
        }
        const promises = this.endpoints.map((ep) => {
            return new Promise((resolve, reject) => {
                /* c8 ignore next */
                if (doDebug) {
                    debugLog("Suspending ", ep.endpointDescriptions()[0].endpointUrl);
                }
                ep.suspendConnection((err) => {
                    /* c8 ignore next */
                    if (doDebug) {
                        debugLog("Suspended ", ep.endpointDescriptions()[0].endpointUrl);
                    }
                    err ? reject(err) : resolve();
                });
            });
        });
        Promise.all(promises)
            .then(() => callback())
            .catch((err) => callback(err));
    }
    resumeEndPoints(callback) {
        // c8 ignore next
        if (!callback)
            throw new Error("thenify is not available");
        const promises = this.endpoints.map((ep) => {
            return new Promise((resolve, reject) => {
                ep.restoreConnection((err) => (err ? reject(err) : resolve()));
            });
        });
        Promise.all(promises)
            .then(() => callback())
            .catch((err) => callback(err));
    }
    prepare(_message, _channel) {
        /* empty */
    }
    /**
     * @private
     */
    _on_GetEndpointsRequest(message, channel) {
        const request = message.request;
        (0, node_opcua_assert_1.assert)(request.schema.name === "GetEndpointsRequest");
        const response = new node_opcua_service_endpoints_1.GetEndpointsResponse({});
        /**
         * endpointUrl	String	The network address that the Client used to access the DiscoveryEndpoint.
         *                      The Server uses this information for diagnostics and to determine what URLs to return in the response.
         *                      The Server should return a suitable default URL if it does not recognize the HostName in the URL
         * localeIds   []LocaleId	List of locales to use.
         *                          Specifies the locale to use when returning human readable strings.
         * profileUris []	String	List of Transport Profile that the returned Endpoints shall support.
         *                          OPC 10000-7 defines URIs for the Transport Profiles.
         *                          All Endpoints are returned if the list is empty.
         *                          If the URI is a URL, this URL may have a query string appended.
         *                          The Transport Profiles that support query strings are defined in OPC 10000-7.
         */
        response.endpoints = this.findMatchingEndpoints(null);
        const _e = response.endpoints.map((e) => e.endpointUrl);
        if (request.endpointUrl) {
            const filtered = response.endpoints.filter((endpoint) => (0, node_opcua_utils_1.matchUri)(endpoint.endpointUrl, request.endpointUrl));
            if (filtered.length > 0) {
                response.endpoints = filtered;
            }
        }
        response.endpoints = response.endpoints.filter((endpoint) => !endpoint.restricted);
        // apply filters
        if (request.profileUris && request.profileUris.length > 0) {
            const profileUris = request.profileUris;
            response.endpoints = response.endpoints.filter((endpoint) => profileUris.indexOf(endpoint.transportProfileUri) >= 0);
        }
        // adjust locale on ApplicationName to match requested local or provide
        // a string with neutral locale (locale === null)
        // TODO: find a better way to handle this
        response.endpoints.forEach((endpoint) => {
            endpoint.server.applicationName.locale = "en-US";
        });
        channel.send_response("MSG", response, message, emptyCallback);
    }
    /**
     * @private
     */
    _on_FindServersRequest(message, channel) {
        // Release 1.02  13  OPC Unified Architecture, Part 4 :
        //   This  Service  can be used without security and it is therefore vulnerable to Denial Of Service (DOS)
        //   attacks. A  Server  should minimize the amount of processing required to send the response for this
        //   Service.  This can be achieved by preparing the result in advance.   The  Server  should  also add a
        //   short delay before starting processing of a request during high traffic conditions.
        const shortDelay = 100; // milliseconds
        setTimeout(() => {
            const request = message.request;
            (0, node_opcua_assert_1.assert)(request.schema.name === "FindServersRequest");
            if (!(request instanceof node_opcua_service_discovery_1.FindServersRequest)) {
                throw new Error("Invalid request type");
            }
            let servers = this.getServers(channel);
            // apply filters
            // TODO /
            if (request.serverUris && request.serverUris.length > 0) {
                const serverUris = request.serverUris;
                // A serverUri matches the applicationUri from the ApplicationDescription define
                servers = servers.filter((inner_Server) => {
                    return serverUris.indexOf(inner_Server.applicationUri) >= 0;
                });
            }
            function adapt(applicationDescription) {
                return new node_opcua_service_endpoints_1.ApplicationDescription({
                    applicationName: applicationDescription.applicationName,
                    applicationType: applicationDescription.applicationType,
                    applicationUri: applicationDescription.applicationUri,
                    discoveryProfileUri: applicationDescription.discoveryProfileUri,
                    discoveryUrls: applicationDescription.discoveryUrls,
                    gatewayServerUri: applicationDescription.gatewayServerUri,
                    productUri: applicationDescription.productUri
                });
            }
            const response = new node_opcua_service_discovery_1.FindServersResponse({
                servers: servers.map(adapt)
            });
            channel.send_response("MSG", response, message, emptyCallback);
        }, shortDelay);
    }
    /**
     * returns a array of currently active channels
     */
    getChannels() {
        let channels = [];
        for (const endpoint of this.endpoints) {
            const c = endpoint.getChannels();
            channels = channels.concat(c);
        }
        return channels;
    }
}
exports.OPCUABaseServer = OPCUABaseServer;
/**
 * construct a service Fault response
 */
function makeServiceFault(statusCode, messages) {
    const response = new node_opcua_service_secure_channel_1.ServiceFault();
    response.responseHeader.serviceResult = statusCode;
    // xx response.serviceDiagnostics.push( new DiagnosticInfo({ additionalInfo: messages.join("\n")}));
    (0, node_opcua_assert_1.assert)(Array.isArray(messages));
    (0, node_opcua_assert_1.assert)(typeof messages[0] === "string");
    response.responseHeader.stringTable = messages;
    // tslint:disable:no-console
    warningLog(chalk_1.default.cyan(" messages "), messages.join("\n"));
    return response;
}
// tslint:disable:no-var-requires
const thenify_ex_1 = require("thenify-ex");
const opts = { multiArgs: false };
OPCUABaseServer.prototype.resumeEndPoints = (0, thenify_ex_1.withCallback)(OPCUABaseServer.prototype.resumeEndPoints, opts);
OPCUABaseServer.prototype.suspendEndPoints = (0, thenify_ex_1.withCallback)(OPCUABaseServer.prototype.suspendEndPoints, opts);
OPCUABaseServer.prototype.shutdownChannels = (0, thenify_ex_1.withCallback)(OPCUABaseServer.prototype.shutdownChannels, opts);
