// node-opcua-common/dist/opcua_secure_object.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPCUASecureObject = exports.SecretHolder = void 0;
exports.invalidateCachedSecrets = invalidateCachedSecrets;
exports.getPartialCertificateChain = getPartialCertificateChain;
/**
 * @module node-opcua-common
 */
const node_events_1 = require("node:events");
const node_fs_1 = __importDefault(require("node:fs"));
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_crypto_1 = require("node-opcua-crypto");
const web_1 = require("node-opcua-crypto/web");
/**
 * Holds cryptographic secrets (certificate chain and private key) for a
 * certificate/key file pair. Secrets are lazily loaded from disk on first
 * access and kept in truly private `#`-fields so they never appear in
 * `JSON.stringify`, `console.log`, `Object.keys`, or `util.inspect`.
 */
class SecretHolder {
    #certificateChain = null;
    #privateKey = null;
    #obj;
    constructor(obj) {
        this.#obj = obj;
    }
    getCertificate() {
        // Ensure the chain is loaded before accessing [0]
        const chain = this.getCertificateChain();
        return chain[0];
    }
    getCertificateChain() {
        if (!this.#certificateChain) {
            const file = this.#obj.certificateFile;
            if (!node_fs_1.default.existsSync(file)) {
                throw new Error(`Certificate file must exist: ${file}`);
            }
            const chain = (0, node_opcua_crypto_1.readCertificateChain)(file);
            if (!chain || chain.length === 0) {
                throw new Error(`Invalid certificate chain (length=0) ${file}`);
            }
            this.#certificateChain = chain;
        }
        return this.#certificateChain;
    }
    getPrivateKey() {
        if (!this.#privateKey) {
            const file = this.#obj.privateKeyFile;
            if (!node_fs_1.default.existsSync(file)) {
                throw new Error(`Private key file must exist: ${file}`);
            }
            const key = (0, node_opcua_crypto_1.readPrivateKey)(file);
            if (key instanceof Buffer) {
                throw new Error(`Invalid private key ${file}. Should not be a buffer`);
            }
            this.#privateKey = key;
        }
        return this.#privateKey;
    }
    /**
     * Clears cached secrets so the GC can reclaim sensitive material.
     * After calling dispose the holder will re-read from disk on next access.
     */
    dispose() {
        this.#certificateChain = null;
        this.#privateKey = null;
    }
    /**
     * Alias for {@link dispose}.
     * Implements `ICertificateChainProvider.invalidate()`.
     */
    invalidate() {
        this.dispose();
    }
    // Prevent secrets from leaking through JSON serialization
    toJSON() {
        return { certificateFile: this.#obj.certificateFile, privateKeyFile: this.#obj.privateKeyFile };
    }
    // Prevent secrets from leaking through console.log / util.inspect
    [Symbol.for("nodejs.util.inspect.custom")]() {
        return `SecretHolder { certificateFile: "${this.#obj.certificateFile}", privateKeyFile: "${this.#obj.privateKeyFile}" }`;
    }
}
exports.SecretHolder = SecretHolder;
/**
 * Module-private WeakMap that associates an ICertificateKeyPairProvider
 * with its SecretHolder. Using a WeakMap means:
 * - The secret holder is invisible from the outside (no enumerable property)
 * - If the owning object is GC'd, the SecretHolder is automatically collected
 */
const secretHolders = new WeakMap();
function getSecretHolder(obj) {
    let holder = secretHolders.get(obj);
    if (!holder) {
        holder = new SecretHolder(obj);
        secretHolders.set(obj, holder);
    }
    return holder;
}
/**
 * Invalidate any cached certificate chain and private key for the given
 * provider so that the next `getCertificate()` / `getPrivateKey()` call
 * re-reads from disk.
 *
 * This is the public replacement for the old `$$certificateChain = null`
 * / `$$privateKey = null` pattern.
 */
function invalidateCachedSecrets(obj) {
    const holder = secretHolders.get(obj);
    if (holder) {
        holder.dispose();
    }
}
/**
 * Extract a partial certificate chain from a certificate chain so that the
 * total size of the chain does not exceed maxSize.
 * If maxSize is not provided, the full certificate chain is returned.
 * If the first certificate in the chain already exceeds maxSize, an error is thrown.
 *
 * @param certificateChain - full certificate chain (single DER buffer or array)
 * @param maxSize          - optional byte budget
 * @returns the truncated chain as an array of individual certificates
 */
function getPartialCertificateChain(certificateChain, maxSize) {
    if (!certificateChain ||
        (Array.isArray(certificateChain) && certificateChain.length === 0) ||
        (certificateChain instanceof Buffer && certificateChain.length === 0)) {
        return [];
    }
    const certificates = Array.isArray(certificateChain) ? certificateChain : (0, web_1.split_der)(certificateChain);
    if (maxSize === undefined) {
        return certificates;
    }
    // at least include first certificate
    const chainToReturn = [certificates[0]];
    let cumulatedLength = certificates[0].length;
    // Throw if first certificate already exceed maxSize
    if (cumulatedLength > maxSize) {
        throw new Error(`getPartialCertificateChain not enough space for leaf certificate ${maxSize} < ${cumulatedLength}`);
    }
    let index = 1;
    while (index < certificates.length && cumulatedLength + certificates[index].length <= maxSize) {
        chainToReturn.push(certificates[index]);
        cumulatedLength += certificates[index].length;
        index++;
    }
    return chainToReturn;
}
/**
 * An object that provides a certificate and a privateKey.
 * Secrets are loaded lazily and stored in a module-private WeakMap
 * so they never appear on the instance.
 */
// biome-ignore lint/suspicious/noExplicitAny: EventEmitter use any
class OPCUASecureObject extends node_events_1.EventEmitter {
    certificateFile;
    privateKeyFile;
    constructor(options) {
        super();
        (0, node_opcua_assert_1.assert)(typeof options.certificateFile === "string");
        (0, node_opcua_assert_1.assert)(typeof options.privateKeyFile === "string");
        this.certificateFile = options.certificateFile || "invalid certificate file";
        this.privateKeyFile = options.privateKeyFile || "invalid private key file";
    }
    getCertificate() {
        return getSecretHolder(this).getCertificate();
    }
    getCertificateChain() {
        return getSecretHolder(this).getCertificateChain();
    }
    getPrivateKey() {
        return getSecretHolder(this).getPrivateKey();
    }
}
exports.OPCUASecureObject = OPCUASecureObject;
