// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/access_right_variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAccessRightVariables = addAccessRightVariables;
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_variant_1 = require("node-opcua-variant");
const defaultInt32Value = new node_opcua_variant_1.Variant({
    dataType: node_opcua_variant_1.DataType.Int32,
    arrayType: node_opcua_variant_1.VariantArrayType.Scalar,
    value: 36
});
const accessRightConfigs = [
    { name: "AccessLevel_CurrentRead", accessLevel: "CurrentRead", userAccessLevel: "CurrentRead" },
    { name: "AccessLevel_CurrentWrite", accessLevel: "CurrentWrite", userAccessLevel: "CurrentWrite", value: {} },
    { name: "AccessLevel_CurrentRead_NotUser", accessLevel: "CurrentRead", userAccessLevel: "" },
    { name: "AccessLevel_CurrentWrite_NotUser", accessLevel: "CurrentWrite | CurrentRead", userAccessLevel: "CurrentRead" },
    { name: "AccessLevel_CurrentRead_NotCurrentWrite", accessLevel: "CurrentRead", userAccessLevel: "CurrentRead" },
    { name: "AccessLevel_CurrentWrite_NotCurrentRead", accessLevel: "CurrentWrite", userAccessLevel: "CurrentWrite" },
    { name: "AccessLevel_DeniedAll", accessLevel: "", userAccessLevel: "" }
];
function addAccessRightVariables(namespace, parentFolder) {
    const accessRight_Folder = namespace.addFolder(parentFolder, {
        browseName: "AccessRight",
        description: "Folder containing various nodes with different access right behavior",
        nodeId: "s=AccessRight"
    });
    const accessLevel_All_Folder = namespace.addFolder(accessRight_Folder, {
        browseName: "AccessLevel",
        description: "Various node with different access right behavior",
        nodeId: "s=AccessLevel"
    });
    for (const config of accessRightConfigs) {
        namespace.addVariable({
            componentOf: accessLevel_All_Folder,
            browseName: config.name,
            description: { locale: "en", text: config.name },
            nodeId: `s=${config.name}`,
            dataType: "Int32",
            valueRank: -1,
            accessLevel: (0, node_opcua_data_model_1.makeAccessLevelFlag)(config.accessLevel),
            userAccessLevel: (0, node_opcua_data_model_1.makeAccessLevelFlag)(config.userAccessLevel),
            value: config.value ?? defaultInt32Value
        });
    }
}
