// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/analog_data_items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAnalogDataItems = addAnalogDataItems;
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_data_access_1 = require("node-opcua-data-access");
const node_opcua_variant_1 = require("node-opcua-variant");
function makeRange(dataType) {
    let engineeringUnitsRange = { low: -200, high: 200 };
    let instrumentRange = { low: -200, high: 200 };
    if (node_opcua_variant_1.DataType[dataType][0] === "U" || dataType === node_opcua_variant_1.DataType.Byte) {
        engineeringUnitsRange = { low: 10, high: 250 };
        instrumentRange = { low: 10, high: 250 };
    }
    return { engineeringUnitsRange, instrumentRange };
}
function addAnalogDataItems(namespace, parentFolder) {
    function _addDataItem(localParentFolder, dataType, initialValue) {
        const name = `${node_opcua_variant_1.DataType[dataType]}DataItem`;
        const nodeId = `s=${name}`;
        const v = namespace.addDataItem({
            componentOf: localParentFolder,
            nodeId,
            browseName: name,
            definition: "(tempA -25) + tempB",
            dataType,
            value: {
                arrayType: node_opcua_variant_1.VariantArrayType.Scalar,
                dataType,
                value: initialValue
            }
        });
        (0, node_opcua_assert_1.assert)(v.nodeId.toString() === `ns=${namespace.index};${nodeId}`, `ns=${namespace.index};${nodeId} ${v.nodeId.toString()}`);
    }
    function _addAnalogDataItem(localParentFolder, dataType, initialValue) {
        const { engineeringUnitsRange, instrumentRange } = makeRange(dataType);
        (0, node_opcua_assert_1.assert)(Array.isArray(initialValue) ||
            (initialValue >= engineeringUnitsRange.low && initialValue <= engineeringUnitsRange.high));
        const name = `${node_opcua_variant_1.DataType[dataType]}AnalogDataItem`;
        const nodeId = `s=${name}`;
        namespace.addAnalogDataItem({
            componentOf: localParentFolder,
            nodeId,
            browseName: name,
            definition: "...",
            valuePrecision: 0.5,
            engineeringUnitsRange,
            instrumentRange,
            engineeringUnits: node_opcua_data_access_1.standardUnits.degree_celsius,
            dataType,
            value: { arrayType: node_opcua_variant_1.VariantArrayType.Scalar, dataType, value: initialValue }
        });
    }
    function _addArrayAnalogDataItem(localParentFolder, dataType, initialValue) {
        const name = `${node_opcua_variant_1.DataType[dataType]}ArrayAnalogDataItem`;
        const nodeId = `s=${name}`;
        const { engineeringUnitsRange, instrumentRange } = makeRange(dataType);
        namespace.addAnalogDataItem({
            componentOf: localParentFolder,
            nodeId,
            browseName: name,
            definition: "(tempA -25) + tempB",
            valuePrecision: 0.5,
            engineeringUnitsRange,
            instrumentRange,
            engineeringUnits: node_opcua_data_access_1.standardUnits.degree_celsius,
            dataType,
            value: {
                arrayType: node_opcua_variant_1.VariantArrayType.Array,
                dataType,
                value: [initialValue, initialValue, initialValue, initialValue, initialValue]
            }
        });
    }
    const analogItemFolder = namespace.addObject({
        organizedBy: parentFolder,
        browseName: "Simulation_AnalogDataItem",
        typeDefinition: "FolderType",
        nodeId: "s=Simulation_AnalogDataItem"
    });
    const { engineeringUnitsRange, instrumentRange } = makeRange(node_opcua_variant_1.DataType.Double);
    namespace.addAnalogDataItem({
        componentOf: analogItemFolder,
        nodeId: "s=DoubleAnalogDataItemWithEU",
        browseName: "DoubleAnalogDataItemWithEU",
        definition: "(tempA -25) + tempB",
        valuePrecision: 0.5,
        engineeringUnitsRange,
        instrumentRange,
        engineeringUnits: node_opcua_data_access_1.standardUnits.degree_celsius,
        dataType: node_opcua_variant_1.DataType.Double,
        value: { dataType: node_opcua_variant_1.DataType.Double, value: 19.5 }
    });
    const data = [
        { dataType: node_opcua_variant_1.DataType.Double, value: 3.14 },
        { dataType: node_opcua_variant_1.DataType.Float, value: 3.14 },
        { dataType: node_opcua_variant_1.DataType.Int16, value: -10 },
        { dataType: node_opcua_variant_1.DataType.UInt16, value: 10 },
        { dataType: node_opcua_variant_1.DataType.Int32, value: -100 },
        { dataType: node_opcua_variant_1.DataType.UInt32, value: 100 },
        { dataType: node_opcua_variant_1.DataType.Int64, value: [0, 0] },
        { dataType: node_opcua_variant_1.DataType.UInt64, value: [0, 0] },
        { dataType: node_opcua_variant_1.DataType.Byte, value: 65 },
        { dataType: node_opcua_variant_1.DataType.SByte, value: -23 }
    ];
    for (const e of data) {
        _addAnalogDataItem(analogItemFolder, e.dataType, e.value);
    }
    for (const e of data) {
        _addDataItem(analogItemFolder, e.dataType, e.value);
    }
    _addDataItem(analogItemFolder, node_opcua_variant_1.DataType.String, "some string");
    _addDataItem(analogItemFolder, node_opcua_variant_1.DataType.DateTime, new Date());
    for (const e of data) {
        _addArrayAnalogDataItem(analogItemFolder, e.dataType, e.value);
    }
}
