// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getValidatorFuncForType = getValidatorFuncForType;
exports.getRandomFuncForType = getRandomFuncForType;
exports.findDataType = findDataType;
exports.makeVariant = makeVariant;
exports.addVariable = addVariable;
exports.addScalarVariable = addScalarVariable;
exports.addArrayVariable = addArrayVariable;
exports.addMultiDimensionalArrayVariable = addMultiDimensionalArrayVariable;
const node_opcua_assert_1 = require("node-opcua-assert");
const ec = __importStar(require("node-opcua-basic-types"));
const node_opcua_basic_types_1 = require("node-opcua-basic-types");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_debug_1 = require("node-opcua-debug");
const node_opcua_factory_1 = require("node-opcua-factory");
const node_opcua_variant_1 = require("node-opcua-variant");
const debugLog = (0, node_opcua_debug_1.make_debugLog)(__filename);
const warningLog = (0, node_opcua_debug_1.make_warningLog)(__filename);
// ─── Validators & Random Generators ─────────────────────────────────────
function defaultValidator( /*value*/) {
    return true;
}
function getValidatorFuncForType(dataType) {
    const f = ec[`isValid${dataType}`];
    return f || defaultValidator;
}
function getRandomFuncForType(dataType) {
    const dataTypeName = node_opcua_variant_1.DataType[dataType];
    const f = ec[`random${dataTypeName}`];
    if (f) {
        return f;
    }
    switch (dataTypeName) {
        case "Variant":
            return () => {
                return new node_opcua_variant_1.Variant();
            };
        case "QualifiedName":
            return () => {
                return new node_opcua_data_model_1.QualifiedName({ name: (0, node_opcua_basic_types_1.randomString)() });
            };
        case "LocalizedText":
            return () => {
                return new node_opcua_data_model_1.LocalizedText({ text: (0, node_opcua_basic_types_1.randomString)() });
            };
        case "XmlElement":
            return () => {
                const element = (0, node_opcua_basic_types_1.randomString)();
                const content = (0, node_opcua_basic_types_1.randomString)();
                return `<${element}>${content}</${element}>`;
            };
        default:
            // c8 ignore next
            throw new Error(`Cannot find random${dataTypeName}() func anywhere`);
    }
}
// ─── DataType Lookup ────────────────────────────────────────────────────
function findDataType(dataTypeName) {
    const builtInDataTypeName = (0, node_opcua_factory_1.findBuiltInType)(dataTypeName);
    const dataType = node_opcua_variant_1.DataType[builtInDataTypeName.name];
    // c8 ignore next
    if (!dataType) {
        throw new Error(` dataType ${dataTypeName} must exists`);
    }
    return dataType;
}
// ─── Validation ─────────────────────────────────────────────────────────
function validate_value_or_array(isArray, variantValue, validatorFunc) {
    (0, node_opcua_assert_1.assert)(typeof validatorFunc === "function");
    if (isArray) {
        const arr = variantValue;
        const n = Math.min(10, arr.length);
        for (let i = 0; i < n; i++) {
            // c8 ignore next
            if (!validatorFunc(arr[i])) {
                throw new Error(`default value must be valid for dataType ${variantValue} at index ${i} got ${arr[i]}`);
            }
        }
    }
    else {
        // c8 ignore next
        if (!validatorFunc(variantValue)) {
            throw new Error(`default value must be valid for dataType ${variantValue}`);
        }
    }
}
// ─── Variant Construction ───────────────────────────────────────────────
function makeVariant(dataTypeName, arrayType, dimensions, current_value) {
    const dataType = findDataType(dataTypeName);
    const validatorFunc = getValidatorFuncForType(dataType);
    const isArray = arrayType === node_opcua_variant_1.VariantArrayType.Array || arrayType === node_opcua_variant_1.VariantArrayType.Matrix;
    validate_value_or_array(isArray, current_value, validatorFunc);
    return new node_opcua_variant_1.Variant({
        dataType,
        arrayType,
        dimensions,
        value: current_value
    });
}
// ─── Generic Variable Helpers ───────────────────────────────────────────
function addVariable(namespace, parent, varName, dataTypeName, current_value, valueRank, arrayDimensions, extra_name) {
    (0, node_opcua_assert_1.assert)(typeof extra_name === "string");
    const arrayType = valueRank <= 0 ? node_opcua_variant_1.VariantArrayType.Scalar : valueRank === 1 ? node_opcua_variant_1.VariantArrayType.Array : node_opcua_variant_1.VariantArrayType.Matrix;
    const variant = makeVariant(dataTypeName, arrayType, arrayDimensions, current_value);
    const name = varName + extra_name;
    const nodeId = `${parent.nodeId.toString()}_${varName}${extra_name}`;
    const placeholder = { variant };
    const variable = namespace.addVariable({
        browseName: name,
        componentOf: parent,
        dataType: varName,
        description: { locale: "en", text: name },
        nodeId,
        valueRank,
        arrayDimensions,
        accessLevel: (0, node_opcua_data_model_1.makeAccessLevelFlag)("CurrentRead | CurrentWrite"),
        userAccessLevel: (0, node_opcua_data_model_1.makeAccessLevelFlag)("CurrentRead | CurrentWrite"),
        value: variant
    });
    variable._backdoor_placeholder = placeholder;
    return variable;
}
function addScalarVariable(namespace, parent, name, realType, default_value, extra_name) {
    (0, node_opcua_assert_1.assert)(typeof extra_name === "string");
    const initialValue = typeof default_value === "function" ? default_value() : default_value;
    const variable = addVariable(namespace, parent, name, realType, initialValue, -1, null, extra_name);
    (0, node_opcua_assert_1.assert)(variable.valueRank === -1);
    (0, node_opcua_assert_1.assert)(variable.accessLevel === node_opcua_data_model_1.AccessLevelFlag.CurrentRead + node_opcua_data_model_1.AccessLevelFlag.CurrentWrite);
    (0, node_opcua_assert_1.assert)(variable.userAccessLevel === node_opcua_data_model_1.AccessLevelFlag.CurrentRead + node_opcua_data_model_1.AccessLevelFlag.CurrentWrite);
    (0, node_opcua_assert_1.assert)(variable.historizing === false);
    return variable;
}
function addArrayVariable(namespace, parent, dataTypeName, default_value, realTypeName, arrayLength, extra_name) {
    (0, node_opcua_assert_1.assert)(typeof dataTypeName === "string");
    (0, node_opcua_assert_1.assert)(typeof realTypeName === "string");
    // c8 ignore next
    if (!node_opcua_variant_1.DataType[realTypeName]) {
        warningLog("dataTypeName", dataTypeName);
        warningLog("realTypeName", realTypeName);
    }
    (0, node_opcua_assert_1.assert)(node_opcua_variant_1.DataType[realTypeName], " expecting a valid real type");
    arrayLength = arrayLength || 10;
    const local_defaultValue = typeof default_value === "function" ? default_value() : default_value;
    const current_value = (0, node_opcua_variant_1.buildVariantArray)(node_opcua_variant_1.DataType[realTypeName], arrayLength, local_defaultValue);
    const variable = addVariable(namespace, parent, dataTypeName, realTypeName, current_value, 1, null, extra_name);
    (0, node_opcua_assert_1.assert)(variable.valueRank === 1);
    (0, node_opcua_assert_1.assert)(variable.accessLevel === node_opcua_data_model_1.AccessLevelFlag.CurrentRead + node_opcua_data_model_1.AccessLevelFlag.CurrentWrite);
    (0, node_opcua_assert_1.assert)(variable.userAccessLevel === node_opcua_data_model_1.AccessLevelFlag.CurrentRead + node_opcua_data_model_1.AccessLevelFlag.CurrentWrite);
    (0, node_opcua_assert_1.assert)(variable.historizing === false);
}
function addMultiDimensionalArrayVariable(namespace, parent, dataTypeName, default_value, realTypeName, nbRows, nbCols, extra_name) {
    (0, node_opcua_assert_1.assert)(typeof dataTypeName === "string");
    (0, node_opcua_assert_1.assert)(typeof realTypeName === "string");
    // c8 ignore next
    if (!node_opcua_variant_1.DataType[realTypeName]) {
        debugLog("dataTypeName", dataTypeName);
        debugLog("realTypeName", realTypeName);
    }
    (0, node_opcua_assert_1.assert)(node_opcua_variant_1.DataType[realTypeName], " expecting a valid real type");
    nbRows = nbRows || 6;
    nbCols = nbCols || 2;
    const local_defaultValue = typeof default_value === "function" ? default_value() : default_value;
    const current_value = (0, node_opcua_variant_1.buildVariantArray)(node_opcua_variant_1.DataType[realTypeName], nbRows * nbCols, local_defaultValue);
    const variable = addVariable(namespace, parent, dataTypeName, realTypeName, current_value, 2, [nbRows, nbCols], extra_name);
    (0, node_opcua_assert_1.assert)(variable.valueRank === 2);
    (0, node_opcua_assert_1.assert)(variable.accessLevel === node_opcua_data_model_1.AccessLevelFlag.CurrentRead + node_opcua_data_model_1.AccessLevelFlag.CurrentWrite);
    (0, node_opcua_assert_1.assert)(variable.userAccessLevel === node_opcua_data_model_1.AccessLevelFlag.CurrentRead + node_opcua_data_model_1.AccessLevelFlag.CurrentWrite);
    (0, node_opcua_assert_1.assert)(variable.historizing === false);
}
