// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/mass_variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addMassVariables = addMassVariables;
const helpers_1 = require("./helpers");
const type_defaults_1 = require("./type_defaults");
function addMassVariablesOfType(namespace, parent, dataTypeName, default_value, realType) {
    const nodeName = `Scalar_Mass_${dataTypeName}`;
    const scalarMass_Type = namespace.addObject({
        browseName: nodeName,
        description: "This folder will contain 100 items per supported data-type.",
        nodeId: `s=${nodeName}`,
        organizedBy: parent
    });
    for (let i = 0; i <= 99; i++) {
        const extra_name = `_${i.toString().padStart(2, "0")}`;
        const local_defaultValue = typeof default_value === "function" ? default_value() : default_value;
        (0, helpers_1.addVariable)(namespace, scalarMass_Type, dataTypeName, realType, local_defaultValue, -1, null, extra_name);
    }
}
function addMassVariables(namespace, scalarFolder) {
    const scalarMass = namespace.addFolder(scalarFolder, {
        browseName: "Scalar_Mass",
        description: "This folder will contain 100 items per supported data-type.",
        nodeId: "s=Scalar_Mass"
    });
    for (const e of type_defaults_1.typeAndDefaultValue) {
        const realType = e.realType || e.type;
        addMassVariablesOfType(namespace, scalarMass, e.type, e.defaultValue, realType);
    }
}
