// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/object_with_methods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addObjectWithMethod = addObjectWithMethod;
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const node_opcua_variant_1 = require("node-opcua-variant");
function addObjectWithMethod(namespace, parentFolder) {
    const myObject = namespace.addObject({
        nodeId: "s=ObjectWithMethods",
        organizedBy: parentFolder,
        browseName: "ObjectWithMethods"
    });
    // MethodNoArgs
    const methodNoArgs = namespace.addMethod(myObject, {
        browseName: "MethodNoArgs",
        nodeId: "s=MethodNoArgs"
    });
    (0, node_opcua_assert_1.assert)((0, node_opcua_nodeid_1.makeNodeId)("MethodNoArgs", namespace.index)
        .toString()
        .match(/s=MethodNoArgs/));
    (0, node_opcua_assert_1.assert)(methodNoArgs.nodeId.toString().match(/s=MethodNoArgs/));
    methodNoArgs.bindMethod((_inputArguments, _context, callback) => {
        callback(null, { statusCode: node_opcua_status_code_1.StatusCodes.Good, outputArguments: [] });
    });
    // MethodIO (input + output)
    const methodIO = namespace.addMethod(myObject, {
        browseName: "MethodIO",
        nodeId: (0, node_opcua_nodeid_1.makeNodeId)("MethodIO", namespace.index),
        inputArguments: [
            {
                name: "ShutterLag",
                description: { text: "specifies the number of seconds to wait before the picture is taken " },
                dataType: node_opcua_variant_1.DataType.UInt32
            }
        ],
        outputArguments: [
            {
                name: "Result",
                description: { text: "the result" },
                dataType: "Int32"
            }
        ]
    });
    methodIO.bindMethod((_inputArguments, _context, callback) => {
        callback(null, {
            statusCode: node_opcua_status_code_1.StatusCodes.Good,
            outputArguments: [{ dataType: node_opcua_variant_1.DataType.Int32, value: 42 }]
        });
    });
    // MethodI (input only)
    const methodI = namespace.addMethod(myObject, {
        browseName: "MethodI",
        nodeId: "s=MethodI",
        inputArguments: [
            {
                name: "ShutterLag",
                description: { text: "specifies the number of seconds to wait before the picture is taken " },
                dataType: node_opcua_variant_1.DataType.UInt32
            }
        ]
    });
    methodI.bindMethod((_inputArguments, _context, callback) => {
        callback(null, { statusCode: node_opcua_status_code_1.StatusCodes.Good, outputArguments: [] });
    });
    // MethodO (output only)
    const methodO = namespace.addMethod(myObject, {
        browseName: "MethodO",
        nodeId: "s=MethodO",
        outputArguments: [
            {
                name: "Result",
                description: { text: "the result" },
                dataType: "Int32"
            }
        ]
    });
    methodO.bindMethod((_inputArguments, _context, callback) => {
        callback(null, {
            statusCode: node_opcua_status_code_1.StatusCodes.Good,
            outputArguments: [{ dataType: node_opcua_variant_1.DataType.Int32, value: 42 }]
        });
    });
}
