// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/path_and_large_arrays.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addPath10Deep = addPath10Deep;
exports.addVeryLargeArrayVariables = addVeryLargeArrayVariables;
const helpers_1 = require("./helpers");
const type_defaults_1 = require("./type_defaults");
function addPath10Deep(namespace, simulation_folder) {
    let parent = simulation_folder;
    for (let i = 1; i <= 10; i++) {
        const name = `Path_${i.toString()}Deep`;
        const child = namespace.addObject({
            organizedBy: parent,
            browseName: name,
            description: `A folder at the top of ${i} elements`,
            typeDefinition: "FolderType",
            nodeId: `s=${name}`
        });
        parent = child;
    }
}
function addVeryLargeArrayVariables(namespace, objectsFolder) {
    const scalarStaticLargeArray = namespace.addObject({
        organizedBy: objectsFolder,
        browseName: "Static_Scalar_Large_Array",
        description: "Single dimension, suggested size of 100k-elements per array.",
        nodeId: "s=Static_Scalar_Large_Array"
    });
    for (const e of type_defaults_1.typeAndDefaultValue) {
        const realType = e.realType || e.type;
        (0, helpers_1.addArrayVariable)(namespace, scalarStaticLargeArray, e.type, e.defaultValue, realType, 50 * 1024, "");
    }
}
