// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/simulation_variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addSimulationVariables = addSimulationVariables;
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_basic_types_1 = require("node-opcua-basic-types");
const node_opcua_variant_1 = require("node-opcua-variant");
const helpers_1 = require("./helpers");
const type_defaults_1 = require("./type_defaults");
function addSimulationVariables(namespace, scalarFolder) {
    let values_to_change = [];
    function add_simulation_variable(parent, dataTypeName, defaultValue, realTypeName) {
        realTypeName = realTypeName || dataTypeName;
        const dataType = (0, helpers_1.findDataType)(realTypeName);
        const randomFunc = (0, helpers_1.getRandomFuncForType)(dataType);
        // c8 ignore next
        if (typeof randomFunc !== "function") {
            throw new Error(`a random function must exist for basicType ${dataTypeName}`);
        }
        const variable = (0, helpers_1.addVariable)(namespace, parent, dataTypeName, realTypeName, defaultValue, -1, null, "");
        values_to_change.push({ dataType, randomFunc, variable });
        return variable;
    }
    const simulation = namespace.addObject({
        browseName: "Scalar_Simulation",
        description: "This folder will contain one item per supported data-type.",
        nodeId: "s=Scalar_Simulation",
        organizedBy: scalarFolder
    });
    for (const e of type_defaults_1.typeAndDefaultValue) {
        const dataType = e.type;
        const defaultValue = typeof e.defaultValue === "function" ? e.defaultValue() : e.defaultValue;
        const realType = e.realType || dataType;
        add_simulation_variable(simulation, dataType, defaultValue, realType);
    }
    // Management nodes
    let interval = 2000;
    let enabled = true;
    let timer;
    function change_randomly() {
        for (const element of values_to_change) {
            const variant = element.variable._backdoor_placeholder.variant;
            variant.value = element.randomFunc();
            element.variable.setValueFromSource(variant);
        }
    }
    function delete_Timer() {
        if (timer) {
            try {
                clearInterval(timer);
            }
            catch (_err) {
                /* c8 ignore next */
                // the leak detector may throw if the timer handle is already disposed
            }
            timer = undefined;
        }
    }
    function install_Timer() {
        delete_Timer();
        (0, node_opcua_assert_1.assert)(!timer);
        if (enabled) {
            timer = setInterval(() => {
                change_randomly();
            }, interval);
        }
    }
    function tearDown_Timer() {
        delete_Timer();
        values_to_change = [];
    }
    const intervalVariable = namespace.addVariable({
        browseName: "Interval",
        componentOf: simulation,
        dataType: "UInt16",
        description: { locale: "en", text: "The rate (in msec) of change for all Simulated items." },
        nodeId: "s=Scalar_Simulation_Interval",
        value: new node_opcua_variant_1.Variant({
            arrayType: node_opcua_variant_1.VariantArrayType.Scalar,
            dataType: node_opcua_variant_1.DataType.UInt16,
            value: interval
        })
    });
    intervalVariable.on("value_changed", (dataValue /*,indexRange*/) => {
        const variant = dataValue.value;
        (0, node_opcua_assert_1.assert)(variant instanceof node_opcua_variant_1.Variant);
        (0, node_opcua_assert_1.assert)((0, node_opcua_basic_types_1.isValidUInt16)(variant.value), " value must be valid for dataType");
        interval = variant.value;
        install_Timer();
    });
    const enabledVariable = namespace.addVariable({
        componentOf: simulation,
        browseName: "Enabled",
        description: { locale: "en", text: "Enabled" },
        nodeId: "s=Scalar_Simulation_Enabled",
        dataType: "Boolean",
        value: new node_opcua_variant_1.Variant({
            dataType: node_opcua_variant_1.DataType.Boolean,
            arrayType: node_opcua_variant_1.VariantArrayType.Scalar,
            value: enabled
        })
    });
    enabledVariable.on("value_changed", (dataValue /*,indexRange*/) => {
        const variant = dataValue.value;
        (0, node_opcua_assert_1.assert)(variant instanceof node_opcua_variant_1.Variant);
        (0, node_opcua_assert_1.assert)((0, node_opcua_basic_types_1.isValidBoolean)(variant.value), " value must be valid for dataType");
        enabled = variant.value;
        install_Timer();
    });
    install_Timer();
    const addressSpace = namespace.addressSpace;
    addressSpace.registerShutdownTask(tearDown_Timer);
}
