// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/static_variables.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addStaticVariables = addStaticVariables;
/**
 * Static scalar, array, and multi-dimensional array variables + image loading.
 */
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const node_opcua_debug_1 = require("node-opcua-debug");
const node_opcua_variant_1 = require("node-opcua-variant");
const helpers_1 = require("./helpers");
const type_defaults_1 = require("./type_defaults");
const errorLog = (0, node_opcua_debug_1.make_errorLog)(__filename);
async function addStaticVariables(namespace, scalarFolder) {
    const staticScalarFolder = namespace.addObject({
        organizedBy: scalarFolder,
        browseName: "Scalars",
        description: "This folder will contain one item per supported data-type.",
        nodeId: "s=Static_Scalar"
    });
    // Scalar variables
    for (const e of type_defaults_1.typeAndDefaultValue) {
        const dataType = e.type;
        const realType = e.realType || dataType;
        const defaultValue = typeof e.defaultValue === "function" ? e.defaultValue() : e.defaultValue;
        (0, helpers_1.addScalarVariable)(namespace, staticScalarFolder, dataType, realType, defaultValue, "");
    }
    // Load images
    async function setImage(imageType, filename) {
        const fullPath = node_path_1.default.join(__dirname, "../../data", filename);
        const imageNode = namespace.findNode(`s=Static_Scalar_Image${imageType}`);
        try {
            const data = await node_fs_1.default.promises.readFile(fullPath);
            imageNode.setValueFromSource(new node_opcua_variant_1.Variant({ dataType: node_opcua_variant_1.DataType.ByteString, value: data }));
        }
        catch (_err) {
            errorLog("cannot load file =", fullPath);
        }
    }
    await setImage("BMP", "image.bmp");
    await setImage("PNG", "tux.png");
    await setImage("GIF", "gif-anime.gif");
    await setImage("JPG", "tiger.jpg");
    // Array variables
    const staticArraysFolders = namespace.addObject({
        organizedBy: scalarFolder,
        browseName: "Arrays",
        description: "Single dimension, suggested size of 10-elements per array. Unsupported types will be missing from the address-space.",
        nodeId: "s=Static_Array"
    });
    for (const e of type_defaults_1.typeAndDefaultValue) {
        const realType = e.realType || e.type;
        (0, helpers_1.addArrayVariable)(namespace, staticArraysFolders, e.type, e.defaultValue, realType, 10, "");
    }
    // Multi-dimensional array variables
    const staticMultiDimensionalArrays = namespace.addObject({
        organizedBy: scalarFolder,
        browseName: "Multi-Dimensional Arrays",
        description: "Single dimension, suggested size of 10-elements per array. Unsupported types will be missing from the address-space.",
        nodeId: "s=Static_MultiDimensional_Array"
    });
    for (const e of type_defaults_1.typeAndDefaultValue) {
        const realType = e.realType || e.type;
        (0, helpers_1.addMultiDimensionalArrayVariable)(namespace, staticMultiDimensionalArrays, e.type, e.defaultValue, realType, 2, 3, "");
    }
}
