// node-opcua-address-space-for-conformance-testing/dist/conformance_testing/trigger_nodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addTriggerNodes = addTriggerNodes;
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_variant_1 = require("node-opcua-variant");
function addTriggerNodes(namespace, parentFolder) {
    const addressSpace = namespace.addressSpace;
    const myEvtType = namespace.findObjectType("MyEventType") ||
        namespace.addEventType({
            browseName: "MyEventType",
            subtypeOf: "BaseEventType"
        });
    function _add_trigger_node(parent, browseName, nodeId) {
        const triggerNode = namespace.addVariable({
            browseName,
            nodeId,
            eventSourceOf: parent,
            dataType: "Double",
            typeDefinition: (0, node_opcua_nodeid_1.makeNodeId)(68)
        });
        let value = 100.0;
        const getFunc = () => new node_opcua_variant_1.Variant({ dataType: node_opcua_variant_1.DataType.Double, value });
        const setFunc = (variant) => {
            value = variant.value;
            const server = addressSpace.rootFolder.objects.server;
            server.raiseEvent(myEvtType, {
                message: { dataType: node_opcua_variant_1.DataType.LocalizedText, value: { text: "Hello World" } },
                severity: { dataType: node_opcua_variant_1.DataType.UInt32, value: 32 }
            });
        };
        triggerNode.bindVariable({ get: getFunc, set: setFunc });
    }
    const sampleTriggerNode = namespace.addObject({
        browseName: "SampleTriggerNode",
        eventNotifier: 0x1,
        organizedBy: parentFolder
    });
    _add_trigger_node(sampleTriggerNode, "TriggerNode01", "s=TriggerNode01");
    _add_trigger_node(sampleTriggerNode, "TriggerNode02", "s=TriggerNode02");
}
